ryanoasis/nerd-fonts
# Add `space` glyph to Symbols fonts

#### Description

Added the `space` glyph from DejaVu Sans Mono Regular into both Nerd
Font Complete symbol fonts and saved them as TTF.

I used fontforge and generated the font as a `TrueType` with default options.

This could, in theory, fix #275.

#### Requirements / Checklist

  - If any of the scripts were modified they have been tested and execute without error, e.g.:
    - `./font-patcher Inconsolata.otf --fontawesome --octicons --pomicons`
    - `./gotta-patch-em-all-font-patcher\!.sh Hermit`

#### What does this Pull Request (PR) do?

It adds the space character to the symbols fonts files.

#### How should this be manually tested?

On a Linux machine, check this output with the normal symbol fonts installed (and a fontconfig that automatically falls back to the symbols):

```bash
pango-view -t "Playing some  for you right now"
```

![Screenshot showing rendering problem with the space after the symbol character](https://user-images.githubusercontent.com/1599/46373376-2a2c4480-c68e-11e8-9256-ccd3941c317c.png)

Then install this new version of the font, reset the cache and try again:

```bash
cp src/glyphs/Symbols-2048-em\ Nerd\ Font\ Complete.ttf ~/.local/share/fonts
fc-cache -fv
pango-view -t "Playing some  for you right now"
```

![Screenshot showing that the rendering problem has now disappeared](https://user-images.githubusercontent.com/1599/46373390-37e1ca00-c68e-11e8-967d-4b0b9decf5ab.png)

#### Any background context you can provide?

Pango is one of the most common font rendering libraries on Linux. It's used by GTK/GNOME and a lot of standalone apps, like Rofi, Polybar, and a lot of terminals.

Pango looks for fallback fonts when the current font doesn't contain a given glyph, but it won't check spaces or any other very common "invisible" characters with the assumption that all fonts should have spaces in them.
This makes it so that after Pango switches to Nerd Fonts for a glyph, it won't switch back until it gets a non-space character after it, making it impossible to have spaces after a symbol.
I usually try to work around it by wrapping symbols in parenthesis, or adding a colon after them, but in some apps this isn't really possible to do anything about.

**Issues to look at:**

* https://github.com/DaveDavenport/rofi/issues/837
* https://bugzilla.gnome.org/show_bug.cgi?id=775202 (Pango sometimes shows a replacement character for space (U+0020) when a font lacks a space)
* https://bugzilla.gnome.org/show_bug.cgi?id=780068 (Spaces become boxed 0020 around powerline symbols)

**Would fix icon rendering in a lot of programs, including:**

* Kitty terminal
* Polybar
* Rofi
* Almost all GNOME apps
* Some Ubuntu-specific apps

#### What are the relevant tickets (if any)?

* #275 

#### Screenshots (if appropriate or helpful)

See "Any background" above for some examples.

#### Caveats / problems

1. I assume that it was fine to copy the space character from DejaVu Sans Mono Regular. (From a licensing perspective)
2. Running "gotta patch 'em all" for the Hermit font does not lead to any diff in the repo; I assume because the space character is not included in any of the character ranges.
3. I exported the TTF without knowing what I was doing. I picked "TrueType (TTF)" and left everything at their defaults in fontforge.
4. I searched and searched to see if there was any script that generated the TTF files, but could not find any.
5. I tried changing the `.sfd` files too, but they got no meaningful changes ("last modified" and "window size" information only).
6. Trying to diff the TTF files is not easy, but it looks like some metadata got lost.

```
› diff -y <(git show master:./Symbols-1000-em\ Nerd\ Font\ Complete.ttf | xxd) <(xxd Symbols-1000-em\ Nerd\ Font\ Complete.ttf) | tail -n 40
000c0260: 0001 0e79 0001 0004 0000 0002 0000 0001  ...y...... | 000c0260: 7405 7675 656a 7307 7861 6d61 7269 6e0f  t.vuejs.xa
000c0270: 0000 0001 0000 0000 0000 0001 0000 0000  .......... | 000c0270: 7861 6d61 7269 6e2d 6f75 746c 696e 6516  xamarin-ou
000c0280: d5a4 2708 0000 0000 d463 d07e 0000 0000  ..'......c | 000c0280: 796f 7574 7562 652d 6372 6561 746f 722d  youtube-cr
000c0290: d6df 4d79 0001 0000 0000 0002 6663 6d74  ..My...... | 000c0290: 7374 7564 696f 0e79 6f75 7475 6265 2d67  studio.you
000c02a0: 0000 0018 666c 6f67 0000 0130 0001 010f  ....flog.. | 000c02a0: 616d 696e 6700 0000 0000 0001 ffff 0002  aming.....
000c02b0: 5061 7463 6865 6420 7769 7468 2027 4e65  Patched wi | 000c02b0: 0001 0000 000c 0000 0016 001e 0002 0001  ..........
000c02c0: 7264 2046 6f6e 7473 2050 6174 6368 6572  rd Fonts P | 000c02c0: 0001 0e7a 0001 0004 0000 0002 0000 0001  ...z......
000c02d0: 2720 2868 7474 7073 3a2f 2f67 6974 6875  ' (https:/ | 000c02d0: 0000 0001 0000 0000 0000 0001 0000 0000  ..........
000c02e0: 622e 636f 6d2f 7279 616e 6f61 7369 732f  b.com/ryan | 000c02e0: d76f f76b 0000 0000 d463 d07e 0000 0000  .o.k.....c
000c02f0: 6e65 7264 2d66 6f6e 7473 290a 0a2a 2057  nerd-fonts | 000c02f0: d7d9 74de                                ..t.
000c0300: 6562 7369 7465 3a20 6874 7470 733a 2f2f  ebsite: ht <
000c0310: 7777 772e 6e65 7264 666f 6e74 732e 636f  www.nerdfo <
000c0320: 6d0a 2a20 5665 7273 696f 6e3a 2032 2e30  m.* Versio <
000c0330: 2e30 0a2a 2044 6576 656c 6f70 6d65 6e74  .0.* Devel <
000c0340: 2057 6562 7369 7465 3a20 6874 7470 733a   Website:  <
000c0350: 2f2f 6769 7468 7562 2e63 6f6d 2f72 7961  //github.c <
000c0360: 6e6f 6173 6973 2f6e 6572 642d 666f 6e74  noasis/ner <
000c0370: 730a 2a20 4368 616e 6765 6c6f 673a 2068  s.* Change <
000c0380: 7474 7073 3a2f 2f67 6974 6875 622e 636f  ttps://git <
000c0390: 6d2f 7279 616e 6f61 7369 732f 6e65 7264  m/ryanoasi <
000c03a0: 2d66 6f6e 7473 2f62 6c6f 622f 6d61 7374  -fonts/blo <
000c03b0: 6572 2f63 6861 6e67 656c 6f67 2e6d 6400  er/changel <
000c03c0: 0000 0000 0001 010f 5061 7463 6865 6420  ........Pa <
000c03d0: 7769 7468 2027 4e65 7264 2046 6f6e 7473  with 'Nerd <
000c03e0: 2050 6174 6368 6572 2720 2868 7474 7073   Patcher'  <
000c03f0: 3a2f 2f67 6974 6875 622e 636f 6d2f 7279  ://github. <
000c0400: 616e 6f61 7369 732f 6e65 7264 2d66 6f6e  anoasis/ne <
000c0410: 7473 290a 0a2a 2057 6562 7369 7465 3a20  ts)..* Web <
000c0420: 6874 7470 733a 2f2f 7777 772e 6e65 7264  https://ww <
000c0430: 666f 6e74 732e 636f 6d0a 2a20 5665 7273  fonts.com. <
000c0440: 696f 6e3a 2032 2e30 2e30 0a2a 2044 6576  ion: 2.0.0 <
000c0450: 656c 6f70 6d65 6e74 2057 6562 7369 7465  elopment W <
000c0460: 3a20 6874 7470 733a 2f2f 6769 7468 7562  : https:// <
000c0470: 2e63 6f6d 2f72 7961 6e6f 6173 6973 2f6e  .com/ryano <
000c0480: 6572 642d 666f 6e74 730a 2a20 4368 616e  erd-fonts. <
000c0490: 6765 6c6f 673a 2068 7474 7073 3a2f 2f67  gelog: htt <
000c04a0: 6974 6875 622e 636f 6d2f 7279 616e 6f61  ithub.com/ <
000c04b0: 7369 732f 6e65 7264 2d66 6f6e 7473 2f62  sis/nerd-f <
000c04c0: 6c6f 622f 6d61 7374 6572 2f63 6861 6e67  lob/master <
000c04d0: 656c 6f67 2e6d 6400 0000 0000            elog.md... <
```

I think you might want to clean this up somehow; maybe repeat the process yourself..? Alternatively, if I could get some instructions on how to properly just add the glyph to the TTF file without any additional changes I could repeat that process and replace the commits in this PR.
